### PR TITLE
[ERE-1821] Customize RegistrationProfileAdmin class and add filters

### PR DESCRIPTION
**Pull request by Lexiel46**

### CODE REVIEW

File: rdrf/rdrf/admin.py [hunk 1]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+
+    def resend_activation_email(self, request, queryset):
+        site = get_current_site(request)
+        for profile in queryset:
+            user = profile.user
+            resend_activation_mail(user.email, site, request)
+            user.date_joined = datetime.datetime.now()
+            user.save()
+
+    resend_activation_email.short_description = _('Re-send activation emails')
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Already defined on the superclass `RegistrationAdmin`, so you shouldn't need to duplicate it.

File: rdrf/rdrf/admin.py [hunk 2]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+    def lookups(self, request, model_admin):
+        return [
+            ('True', _('True')),
+            ('False', _('False')),
+        ]
+
+    def queryset(self, request, queryset):
+        expired_emails = []
+        unexpired_emails = []
+        for profile in queryset:
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Should be chaining ORM queries to the existing queryset rather than using a `for` loop.

File: rdrf/rdrf/admin.py [hunk 3]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+            user_email = profile.user.email
+            if profile.activation_key_expired():
+                expired_emails.append(user_email)
+            else:
+                unexpired_emails.append(user_email)
+
+        if self.value() == 'True':
+            return RegistrationProfile.objects.filter(user__email__in=expired_emails)
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: chain to existing `queryset` rather than creating a new filter from the `RegistrationProfile`.

File: rdrf/rdrf/admin.py [hunk 4]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+        if self.value() == 'False':
+            return RegistrationProfile.objects.filter(user__email__in=unexpired_emails)
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Should return the `queryset` if none of these conditions are met to avoid errors.

Could us `if... elif... else` for the various branches expected.

File: rdrf/rdrf/admin.py [hunk 5]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
         return () if obj is None else ("code",)
 
 
+class ActivationKeyExpirationListFilter(admin.SimpleListFilter):
+    title = _('activation key expired')
+    parameter_name = 'activation_key_expired'
+
+    def lookups(self, request, model_admin):
+        return [
+            ('True', _('True')),
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Yes / No is generally used on the UI for True/False.

File: rdrf/rdrf/admin.py [hunk 6]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+            else:
+                unexpired_emails.append(user_email)
+
+        if self.value() == 'True':
+            return RegistrationProfile.objects.filter(user__email__in=expired_emails)
+        if self.value() == 'False':
+            return RegistrationProfile.objects.filter(user__email__in=unexpired_emails)
+
+
+def resend_activation_mail(email, site, request=None):
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: If you pass the user object, you can get the registration profile with `user.registrationprofile`, rather than performing a lookup by email address. You will need to make sure the user has the registrationprofile attribute, because not all users will have these (although the ones that are displayed on this page probably all do)

File: rdrf/rdrf/admin.py [hunk 7]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+    if user_profile.activated:
+        return False
+
+    user_profile.create_new_activation_key()
+    user_profile.send_activation_email(site, request)
+
+    return True
+
+
+def activate_user(activation_key):
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: @ScriptSmith Can you think of why we should do a full activation on the user's profile rather than just updating the user active status to 'Active'?

File: rdrf/rdrf/admin.py [hunk 8]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+    sha256_re = re.compile('^[a-f0-9]{40,64}$')
+
+    def activate(user_profile):
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: I'm concerned about these functions being `public`, can we nest them inside the class they are used or at least rename with an underscore e.g. `_activate` so it is clear the intention is not to use them outside of this admin script?

File: rdrf/rdrf/admin.py [hunk 9]
```diff
@@ -178,6 +183,93 @@ def get_readonly_fields(self, request, obj=None):
...
+        for profile in queryset:
+            activate_user(profile.activation_key)
+    activate_users.short_description = _('Activate users')
+
+    def resend_activation_email(self, request, queryset):
+        site = get_current_site(request)
+        for profile in queryset:
+            user = profile.user
+            resend_activation_mail(user.email, site, request)
+            user.date_joined = datetime.datetime.now()
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: I understand why you've reset the user's date_joined, but I don't know if we should be doing this.

**Review by ppettitau — CHANGES_REQUESTED** [review]

File: rdrf/rdrf/admin.py (hunk 2 above)
Comment (ppettitau) [reply rdrf/rdrf/admin.py]: As discussed, my suggestion to chain ORM queries doesn't work here because you need to call a function on the profile `activation_key_expired()`. 

You  could look at using List Comprehensions to reduce code.

Something like:
```
expired_profiles = [profile.id for profile in queryset if profile.activation_key_expired()]
        
queryset.objects.filter(id__in=[expired_profiles]) 
queryset.objects.exclude(id__in=[expired_profiles])
```


https://realpython.com/list-comprehension-python/#using-list-comprehensions
https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions

File: rdrf/rdrf/admin.py [hunk 10]
```diff
@@ -178,6 +184,83 @@ def get_readonly_fields(self, request, obj=None):
...
+        if self.value() == 'True':
+            return queryset.filter(id__in=expired_profiles)
+        elif self.value() == 'False':
+            return queryset.exclude(id__in=expired_profiles)
+        else:
+            return queryset
+
+
+def resend_activation_mail(profile, site, request=None):
+    try:
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Can you use `hasattr` to check rather than relying on an exception?

File: rdrf/rdrf/admin.py [hunk 11]
```diff
@@ -178,6 +183,87 @@ def get_readonly_fields(self, request, obj=None):
...
+            return queryset.exclude(id__in=expired_profiles)
+        else:
+            return queryset
+
+
+def resend_activation_mail(profile, site, request=None):
+    if not hasattr(profile.user, 'registrationprofile') or profile.activated:
+        return False
+
+    profile.create_new_activation_key()
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: This new activation key doesn't work for the user if their previous activation key has expired.

The original registration model logic to `activate_user` will check if the `activation_key_expired` is true, and if so will not allow the profile to be activated. So if the user you 'resend' the activation email to has joined more than `settings.ACCOUNT_ACTIVATION_DAYS` ago, they will not be able to successfully activate their account.

File: rdrf/rdrf/admin.py [hunk 12]
```diff
@@ -178,6 +183,87 @@ def get_readonly_fields(self, request, obj=None):
...
+        return False, False
+
+    def activate_users(self, request, queryset):
+        for profile in queryset:
+            self.activate_user(profile.activation_key)
+
+    def resend_activation_email(self, request, queryset):
+        site = get_current_site(request)
+        for profile in queryset:
+            resend_activation_mail(profile, site, request)
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: A couple of things with this `resend_activation_mail` function:
* It's only used once (here), so is there any benefit to defining it at the global scope rather than just inlining the code here?
* If you feel there is a benefit to defining it separately, can we rename the methods - I know they are slightly different (email vs mail), but can be easily be mistaken for the same name and result in confusion. Otherwise if we inline it as described above, then we don't have to think of another name :)
* The global function `resend_activation_mail` also returns True/False which isn't checked here

File: rdrf/rdrf/admin.py [hunk 13]
```diff
@@ -178,6 +183,87 @@ def get_readonly_fields(self, request, obj=None):
...
+
+class CustomRegistrationProfileAdmin(RegistrationAdmin):
+    list_display = ('user', 'custom_activation_key_expired', 'activated')
+    list_filter = ['activated', ActivationKeyExpirationListFilter]
+
+    @admin.display(description="Activation key expired")
+    def custom_activation_key_expired(self, obj):
+        max_expiry_days = datetime.timedelta(days=settings.ACCOUNT_ACTIVATION_DAYS)
+        activation_date = obj.user.date_activated
+        expiration_date = (obj.user.date_joined if activation_date is None else activation_date) + max_expiry_days
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: I think you can do:
`(activation_date or obj.user.date_joined) + max_expiry_days` to achieve the same logic

**Review by ppettitau — CHANGES_REQUESTED** [review]

**Comment by ppettitau** [comment]
Icons used to convey information that is not available in text is not recommended for accessibility. Probably a good idea to replace these ticks/crosses with True/False or Yes/No, similar to the 'Activation key expired' column.
image

File: rdrf/rdrf/admin.py [hunk 14]
```diff
@@ -255,11 +250,19 @@ def activate_users(self, request, queryset):
         for profile in queryset:
             self.activate_user(profile.activation_key)
 
+    def get_registration_activation_url(self, registration_profile):
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Why is this required?

File: rdrf/rdrf/views/registration_rdrf.py [hunk 15]
```diff
@@ -144,6 +148,22 @@ def get_success_url(self, user=None):
 
 
 class PatientActivationView(ActivationView):
+
+    def custom_activation_key_expired(self, profile):
```
Comment (ppettitau) [inline rdrf/rdrf/views/registration_rdrf.py]: Is this required here?

File: rdrf/rdrf/admin.py [hunk 16]
```diff
@@ -178,6 +184,89 @@ def get_readonly_fields(self, request, obj=None):
...
+            user_profile.activated = True
+
+            with transaction.atomic():
+                user.save()
+                user_profile.save()
+
+            return user
+
+        if sha256_re.search(activation_key):
+            profile = RegistrationProfile.objects.filter(activation_key=activation_key)
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Should be called `profiles` to indicate it's more than one.

File: rdrf/rdrf/admin.py [hunk 17]
```diff
@@ -178,6 +184,89 @@ def get_readonly_fields(self, request, obj=None):
...
+            if not profile.exists():
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Can remove a lot of this code by just doing:
```
profile = RegistrationProfile.objects.filter(activation_key=activation_key).first()
if profile and not self.custom_activation_key_expired(profile):
    etc...
```

File: rdrf/rdrf/admin.py [hunk 18]
```diff
@@ -178,6 +184,89 @@ def get_readonly_fields(self, request, obj=None):
...
+                return False, False
+
+            profile = profile.first()
+            if not self.custom_activation_key_expired(profile):
```
Comment (ppettitau) [inline rdrf/rdrf/admin.py]: Do you need to check if `profile.activated` similar to how it's checked on the `RegistrationManager` from the `registration` package?

File: rdrf/rdrf/views/registration_rdrf.py [hunk 19]
```diff
@@ -144,6 +148,22 @@ def get_success_url(self, user=None):
...
+        max_expiry_days = datetime.timedelta(days=settings.ACCOUNT_ACTIVATION_DAYS)
+        activation_date = profile.user.date_activated
+        expiration_date = (activation_date or profile.user.date_joined) + max_expiry_days
+
+        return profile.activated or expiration_date <= datetime.datetime.now()
+
+    def activate(self, *args, **kwargs):
+        activation_key = kwargs.get('activation_key', '')
+        user, activated = CustomRegistrationProfileAdmin.activate_user(self, activation_key)
```
Comment (ppettitau) [inline rdrf/rdrf/views/registration_rdrf.py]: I'm not sure how I feel about having this method in the `CustomRegistrationProfileAdmin` class and calling it from here. Thoughts, @ScriptSmith ?

File: rdrf/rdrf/admin.py (hunk 14 above)
Comment (Lexiel46) [reply rdrf/rdrf/admin.py]: Because the `send_activation_email` function in PatientRegistration has code `self.get_registration_activation_url(registration)`. When we call `send_activation_email` function here, the self points to the CustomRegistrationProfileAdmin class, which then require CustomRegistrationProfileAdmin class has a `get_registration_activation_url` function, otherwise, an error will pop up. Same reason for adding `custom_activation_key_expired` function in PatientActivationView.

Comment (ppettitau) [reply rdrf/rdrf/admin.py]: `PatientRegistration` is a class that should be instantiated, and the `send_activation_email` called on the instantiated object. e.g.
```
            patient_registration = PatientRegistration(request)
            patient_registration.send_activation_email(user.registry_code, user, user.patient)
```

Even so, we shouldn't be naming the registration class directly, as it can be customised per registry. It is generally "loaded" via the `registration_rdrf`'s `load_registration_class`.

At this point I'm wondering how this is going to work for MND where patients have been added by clinicians (i.e. NOT self registered), and also what it will look like for carers.

@ScriptSmith Do you think it's worth taking this solution further to address the above, or do we just call it at this point to say it's too complicated, and remove the option to "resend" the activation email?

File: rdrf/rdrf/admin.py [hunk 20]
```diff
@@ -178,6 +180,37 @@ def get_readonly_fields(self, request, obj=None):
...
+
+
+class CustomRegistrationProfileAdmin(RegistrationAdmin):
+    actions = ['activate_users']
+    list_display = ('user', 'activation_key_expired', 'is_activated')
+    list_filter = [ActivationKeyExpirationListFilter, 'activated']
+
+    @admin.display(description="Activated")
+    def is_activated(self, obj):
+        return obj.activated
```
Comment (ScriptSmith) [inline rdrf/rdrf/admin.py:211]: Can be simplified

```suggestion
    list_display = ('user', 'activation_key_expired', 'activated')
    list_filter = [ActivationKeyExpirationListFilter, 'activated']
```

**Review by ScriptSmith — APPROVED** [review]
Looks good

File: rdrf/rdrf/admin.py (hunk 20 above)
Comment (ScriptSmith) [reply rdrf/rdrf/admin.py:211]: Actually, it makes the table inconsistent as it will display True/False and a Tick/Cross, so probably fine as-is

**Review by ppettitau — APPROVED** [review]
